packages/core/js-client/src/api.ts

```diff
@@ -15,182 +15,187 @@
-  const { args, client: peer, config } = extractFunctionArgs(rawFnArgs, def);
+  const [peer, ...rest] = args;
 
-  return callAquaFunction({
-    args,
-    def,
-    script,
-    config,
-    peer,
-  });
-};
+  if (!(peer instanceof FluencePeer)) {
+    await v5_callFunction([getDefaultPeer(), ...rest], def, script);
+    return;
+  }
 
-/**
- * Convenience function to support Aqua `service` generation backend
- * The compiler only need to generate a call the function and provide the corresponding definitions and the air script
- * @param args - raw arguments passed by user to the generated function
- * TODO: dont forget to add jsdoc for new arg
- * @param def - service definition generated by the Aqua compiler
- */
-export const v5_registerService = (args: unknown[], def: ServiceDef): void => {
-  // TODO: Support this in aqua-to-js package
-  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
-  const service: ServiceImpl = args.pop() as ServiceImpl;
+  const argNames = Object.keys(def.arrow);
+  const schemaArgCount = argNames.length;
 
-  const { peer, serviceId } = extractServiceArgs(args, def.defaultServiceId);
+  type FunctionArg = SimpleTypes | ArrowWithoutCallbacks;
 
-  registerService({
-    def,
-    service,
-    serviceId,
-    peer,
-  });
-};
+  const schemaFunctionArgs: Record<string, FunctionArg> =
+    def.arrow.domain.tag === "nil" ? {} : def.arrow.domain.fields;
 
-function isConfig(arg: unknown): arg is FnConfig {
-  return typeof arg === "object" && arg !== null;
-}
-
-/**
- * Arguments could be passed in one these configurations:
- * [...actualArgs]
- * [peer, ...actualArgs]
- * [...actualArgs, config]
- * [peer, ...actualArgs, config]
- *
- * This function select the appropriate configuration and returns
- * arguments in a structured way of: { peer, config, args }
- */
-function extractFunctionArgs(
-  args: unknown[],
-  def: FunctionCallDef,
-): {
-  client: FluencePeer;
-  config: FnConfig;
-  args: PassedArgs;
-} {
-  const argumentTypes = getArgumentTypes(def);
-  const argumentNames = Object.keys(argumentTypes);
-  const numberOfExpectedArgs = argumentNames.length;
-
-  let peer: FluencePeer;
-  let config: FnConfig;
-
-  if (isFluencePeer(args[0])) {
-    peer = args[0];
-    args = args.slice(1);
-  } else {
-    if (Fluence.defaultClient == null) {
-      throw new Error(
-        "Could not register Aqua service because the client is not initialized. Did you forget to call Fluence.connect()?",
-      );
-    }
+  // if args more than expected in schema (schemaArgCount) then last arg is config
+  const config = schemaArgCount < rest.length ? rest.pop() : undefined;
 
-    peer = Fluence.defaultClient;
+  if (!isAquaConfig(config)) {
+    throw new Error("Config should be object type");
   }
 
-  const maybeConfig = args[numberOfExpectedArgs];
+  const callArgs = Object.fromEntries<JSONValue | ServiceImpl[string]>(
+    rest.slice(0, schemaArgCount).map((arg, i) => {
```

Comment:
I would refactor this lambda a little bit
1. add `const argName = argNames[i]` and use it everywhere in the lambda
2. Improve errors, e.g. `Expected argument ${i} to be a function. Found ${typeof arg}: ${arg}`
3. I would inline `wrapFunction(arg, argSchema)` - no need for a variable there in my opinion